### PR TITLE
Add automatic og:image generation to docs

### PR DESCRIPTION
**Pull request by xnameTM**
## Summary
This pr includes changes to dynamically build OG Images for every single subpage in docs.

## How does it work?
The algorithm runs after docusaurus build. It reads titles from markdown files in docs dictionary, generates images and place .png files in the build dictionary.
Docusaurus has been configured to automatically link to generated OG Images.

## How to test this feature?
Command to change current working dictionary:
> cd docs

Command to download libraries:
> yarn install

Command to build project and generete OG Images:
> yarn build

Command to see the result:
> yarn serve

### CODE REVIEW

File: docs/build-og-images.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          lines.push({
+            line,
+            y,
+          });
+          lines.forEach(({ line, y }, index) => {
+            ctx.fillText(line, 67, y - (lines.length - index) * 37);
+          });
+        }
+
+        ctx.font = 'bold 40px "Aeonik Bold"';
```
Comment (patrycjakalinska) [inline docs/build-og-images.js:94]:
```suggestion
        ctx.font = 'normal 40px "Aeonik Regular"';
```
Without bold subtitle it looks slightly better

Before:
image

After:
image

File: docs/build-og-images.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      return {
+        dir,
+        files: files.filter(
+          (file) => file.endsWith('.md') || file.endsWith('.mdx')
+        ),
+      };
+    })
+  );
+
+  const targetDocs = path.resolve(__dirname, 'build/docs/og');
```
Comment (patrycjakalinska) [inline docs/build-og-images.js:39]:
```suggestion
  const targetDocs = path.resolve(__dirname, 'build/img/og');
```
i would keep the images in img folder

**Review by patrycjakalinska — APPROVED** [review]
🚀

File: docs/build-og-images.js
Comment (kacperkapusciak) [inline docs/build-og-images.js:1]: please move the file to `scripts` folder

File: docs/build-og-images.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    .filter((line) => line.startsWith('#'))
+    .map((line, index) => ({
+      level: line.match(/#/g).length,
+      title: line.replace(/#/g, '').trim(),
+      index,
+    }))
+    .sort((a, b) => a.level - b.level || a.index - b.index);
+
+  return headers[0]?.title || '';
+};
```
Comment (kacperkapusciak) [inline docs/build-og-images.js:21]: it might be simpler to just find a line that starts with a single #

File: docs/build-og-images.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  const ctx = canvas.getContext('2d');
+
+  await Promise.all(
+    docs.map(async ({ dir, files }) => {
+      files.map(async (file) => {
+        ctx.drawImage(image, 0, 0, 1200, 630);
+
+        let header = getMarkdownHeader(path.resolve(baseDocsPath, dir, file));
+
+        if (header === '') header = 'unnamed';
```
Comment (kacperkapusciak) [inline docs/build-og-images.js:60]:
```suggestion
        if (header === '') header = 'React Native Reanimated';
```
unnamed isn't a great default title in an OG image, let's use React Native Reanimated instead

File: docs/build-og-images.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        ctx.font = 'bold 72px "Aeonik Bold"';
+        ctx.fillStyle = '#001A72';
+        ctx.textAlign = 'left';
```
Comment (kacperkapusciak) [inline docs/build-og-images.js:64]: I'd extract all the [magic numbers and magic strings](https://en.wikipedia.org/wiki/Magic_number_(programming)) to constants on top of the file

`const TITLE_FONT = 'bold 72px "Aeonik Bold"';`

etc

File: docs/src/theme/DocItem/Metadata/index.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import { useDoc } from '@docusaurus/theme-common/internal';
+import useBaseUrl from '@docusaurus/useBaseUrl';
+export default function DocItemMetadata() {
+  const { metadata, frontMatter, assets } = useDoc();
+
+  if (!metadata.title) {
+    return null;
+  }
+
+  const image = metadata.title
```
Comment (kacperkapusciak) [inline docs/src/theme/DocItem/Metadata/index.js:12]:
```suggestion
  const ogImageName = metadata.title
```
a bit more descriptive name. `image` suggests an object that holds the image but here it's just the name

File: docs/build-og-images.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+              line = newLine;
+            } else {
+              lines.push({
+                line,
+                y,
+              });
+              y += 37;
+              line = `${words[i]}`;
+            }
+          }
```
Comment (kacperkapusciak) [inline docs/build-og-images.js:84]: let's try rewriting this part using `map` function returning a lines array instead of a for loop

**Review by kacperkapusciak — CHANGES_REQUESTED** [review]
I really wish we used https://www.npmjs.com/package/@vercel/og as the declarative approach would be way easier to read and maintain in the long run

File: docs/package.json [hunk 8]
```diff
@@ -32,6 +32,7 @@
     "babel-polyfill": "^6.26.0",
     "babel-preset-expo": "^9.2.2",
     "babel-preset-react-native": "^4.0.1",
+    "canvas": "^2.11.2",
```
Comment (piaskowyk) [inline docs/package.json:35]: Why didn't you try `https://www.npmjs.com/package/@vercel/og`? 🤔

**Review by piaskowyk — CHANGES_REQUESTED** [review]

File: docs/build-og-images.js (hunk 3 above)
Comment (xnameTM) [reply docs/build-og-images.js]: yes, but it take into account situations in which a single # will not be present

File: docs/build-og-images.js (hunk 7 above)
Comment (xnameTM) [reply docs/build-og-images.js]: ok

File: docs/package.json (hunk 8 above)
Comment (xnameTM) [reply docs/package.json:35]: canvas seemed easier

File: docs/scripts/og-image-stream.jsx [hunk 9]
```diff
@@ -0,0 +1,86 @@
+import React from 'react';
+import { ImageResponse } from '@vercel/og';
```
Comment (kacperkapusciak) [inline docs/scripts/og-image-stream.jsx:2]: The code doesn't compile because of a missing `"fs"`import here

**Review by kacperkapusciak — COMMENTED** [review]
One thing is missing before I can approve this. Other than that lgtm

File: docs/scripts/og-image-stream.jsx (hunk 9 above)
Comment (xnameTM) [reply docs/scripts/og-image-stream.jsx:2]: yup

**Review by kacperkapusciak — APPROVED** [review]

**Review by kacperkapusciak — CHANGES_REQUESTED** [review]
overall looks good, are the implementations on Reanimated and RNGH identical? just asking. if not please copy over so they are identical

File: docs/scripts/build-og-images.jsx [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+async function saveStreamToFile(stream, path) {
+  const writeStream = createWriteStream(path);
+  await promisify(pipeline)(stream, writeStream);
+}
+
+const buildOGImages = async () => {
+  const baseDocsPath = path.resolve(__dirname, '../docs');
+  const docs = await Promise.all(
```
Comment (piaskowyk) [inline docs/scripts/build-og-images.jsx:32]: is it `docs` or `directories`?

File: docs/scripts/build-og-images.jsx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      return {
+        dir,
+        files: files.filter(
+          (file) => file.endsWith('.md') || file.endsWith('.mdx')
+        ),
+      };
+    })
+  );
+
+  const targetDocs = path.resolve(__dirname, '../build/img/og');
```
Comment (piaskowyk) [inline docs/scripts/build-og-images.jsx:46]:
```suggestion
  const ogImageTargets = path.resolve(__dirname, '../build/img/og');
```

File: docs/scripts/build-og-images.jsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+  return headers[0]?.title || 'React Native Reanimated';
+};
+
+async function saveStreamToFile(stream, path) {
+  const writeStream = createWriteStream(path);
+  await promisify(pipeline)(stream, writeStream);
+}
+
+const buildOGImages = async () => {
```
Comment (piaskowyk) [inline docs/scripts/build-og-images.jsx:30]:
```suggestion
async function buildOGImages() {
```

File: docs/scripts/build-og-images.jsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        files: files.filter(
+          (file) => file.endsWith('.md') || file.endsWith('.mdx')
+        ),
+      };
+    })
+  );
+
+  const targetDocs = path.resolve(__dirname, '../build/img/og');
+
+  if (fs.existsSync(targetDocs)) fs.rmSync(targetDocs, { recursive: true });
```
Comment (piaskowyk) [inline docs/scripts/build-og-images.jsx:48]:
```suggestion
  if (fs.existsSync(targetDocs)) {
    fs.rmSync(targetDocs, { recursive: true });
  }
```

File: docs/scripts/build-og-images.jsx [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+  fs.mkdirSync(targetDocs, {recursive: true});
```
Comment (piaskowyk) [inline docs/scripts/build-og-images.jsx:50]:
```suggestion
  fs.mkdirSync(targetDocs, { recursive: true });
```

**Review by piaskowyk — APPROVED** [review]

File: docs/scripts/build-og-images.jsx (hunk 10 above)
Comment (xnameTM) [reply docs/scripts/build-og-images.jsx:32]: dirs

File: docs/scripts/build-og-images.jsx (hunk 11 above)
Comment (xnameTM) [reply docs/scripts/build-og-images.jsx:46]: ok

**Review by kacperkapusciak — APPROVED** [review]

**Pull request merged**